### PR TITLE
ENH Adding AUTOSOL file with a package of robust linear equation solvers which detect and handle ill-conditioniong

### PR DESCRIPTION
**Pull request by rejones7**
I am pulling a file that contains two routines, Autosolve(A,b) and Autosolve_nonneg(A,b).
These algorithms have been available for C++ for some years, but have been tweaked for slight performance improvement and converted to Python. Autosolve's advantage over scipy's "solve", or "lstsq", or "lsmr" is that its more universal approach, built on the concept of the Discrete Picard Condition (see referenced in the code) handles every case, whether simply square systems, or underdetermined or overdetermined, and whether consistent or inconsistent, automatically, switching to Tikhonov regularization (with automatically determined parameters) when needed. Also, it needs no extra information like error estimates, tolerances, iteration limits, etc, and there are no algorithmic errors to report. Thus it is ideal for new users, or situations where a system of equations may sometimes be well conditioned and sometimes not. More details in short attached file.
I am quite new to GitHub so please tell me if I am nor doing something right.  Ron Jones rejones7@msn.com,
[submit_autosolve.txt](https://github.com/scipy/scipy/files/5113433/submit_autosolve.txt)

### CODE REVIEW

**Comment by rejones7** [comment]
Ilhan:
I am trying to push (pull??) my new version of autosolve.py, now with test_autosolve.py.
I am so confused with GitHub.
The new code is in my branch:
https://github.com/rejones7/scipy/tree/master/scipy/linalg

But when I try to do a PULL github takes me to the old PULL request, and I don't know how to proceed.

I was a professional mathematician and software developer for decades, and I understand all the code maintence issues, and have taught code development at the university levelm but I just don't understand GitHub and have NO ONE I can get personal help from. I know no one who uses it as a developer.

Can you give an idea what I am supposed to be doing?
Ron Jones
(505)934-8799
[https://avatars1.githubusercontent.com/u/65885180?s=400&v=4]<https://github.com/rejones7/scipy/tree/master/scipy/linalg>
rejones7/scipy<https://github.com/rejones7/scipy/tree/master/scipy/linalg>
Scipy library main repository. Contribute to rejones7/scipy development by creating an account on GitHub.
github.com


________________________________

**Comment by rejones7** [comment]
This is a resubmittal of autosol, which now contains three automatically regularizing routines: 
-autosolve, the primary routine which uses the Picard Condition principle to detect and regularize ill-conditioned (incuding singular) systems. 
-autosolve_nonneg which forces the soltution to be nonnegative;
- autosolve_rising which forces the derivative of the solution to be nonnegative.
Please refer to previous PULL for details, or to http://www.rejones7.net/autorej/index.htm .
A full coverage test is now in the scipy/linalg/tests folder.
A large graphic showing how autosolve compares to lsmr, lstsq, and solve on four test matrices is in scipy/linalg/tests/data. comments invited rejones7@msn.com

**Comment by rejones7** [comment]
A question that came up immediately upon the PULL request is whether I could call a cheaper (non-SVD) solver first before doing the SVD to start my analysis, I have illustrated the problem with that in a chart in scipy/linalg/tests/data/autosolve)comparison. Here is the result: typical solvers such a lstsq cannot see that they are producing garbage when the system is poorly conditioned, though the matrix is nowhere near being singular. There is a fundamental problem with traditional solvers which put the blame for ill-conditioned behavior on the Matrix, and its small singular values. But as we learn when study the Picard Condition, it is the vector S+ * Ut * b
(where Ax=b is being solved, as USVt is the SVD) that tells the real story, and that story is that data error in b is the source of the problem. As the diagram shows, lstsq and solve produce horrible oscillatory solution vectors long before the matrix is singular, and they neither know they are doing this are report it. So my algorithms must stand alone... presolving with lstsq, for example, is useless.

File: scipy/linalg/autosol.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+from numpy import atleast_1d, atleast_2d
+# import scipy.linalg.decomp_svd
+from scipy.linalg.decomp import _asarray_validated
+# from scipy.linalg.misc import LinAlgError, _datacopied, LinAlgWarning
+from scipy.linalg.misc import LinAlgError
+
+# Author : Rondall E. Jones, August 2020
+
+
+def two_norm(x):
```
Comment (ilayn) [inline scipy/linalg/autosol.py:12]: This is basically a rewrite of `scipy.linalg.norm`

File: scipy/linalg/autosol.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+    mx = max(y)
+    if mx == 0.0:
+        return 0.0
+    sm = 0.0
+    for i in range(0, ln):
+        sm += (x[i] / mx) ** 2
+    return mx * sqrt(sm)
+
+
+def rms(x):
```
Comment (ilayn) [inline scipy/linalg/autosol.py:27]: as above you can do `scipy.linalg.norm(a)/len(a)`

File: scipy/linalg/autosol.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    """ Computes root-mean-square of 1-D array x without overflow"""
+    s = two_norm(x)
+    return s / sqrt(len(x))
+
+
+def compute_mov_sums(g, w, m):
```
Comment (ilayn) [inline scipy/linalg/autosol.py:33]: Also here you can use convolution as in `np.convolve(g ,np.ones(w, dtype=int),'valid')` or whichever behavior you prefer (see `numpy.convolve` docstring)

File: scipy/linalg/autosol.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    n = A.shape[1]
+    if b.shape[0] != m:
+        raise LinAlgError("Inconsistent array sizes.")
+    z = np.zeros(n)
+    if np.count_nonzero(A) == 0:
+        return z, 0, 0.0, 0.0
+    if np.count_nonzero(b) == 0:
+        return z, 0, 0.0, 0.0
+
+    C = A  # working copy
```
Comment (ilayn) [inline scipy/linalg/autosol.py:364]: Python doesn't make copies in assignments. You have to make explicit copies via `C = A.copy()`

File: scipy/linalg/tests/test_autosolve.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+# inconsistent arrays .. these crash the run
+# A=np.zeros((3,3))
+# b=np.zeros(5)
+# x=autosolve(A,b)[0]
+# x=autosolve_nonneg(A,b)[0]
+# x=autosolve_rising(A,b)[0]
+
+# 0-D
+print()
```
Comment (ilayn) [inline scipy/linalg/tests/test_autosolve.py:20]: Tests never prints out anything. You can check the other test files for examples for example `linalg/tests/test_basic.py` for inspiration. The results should be asserted and test suite will pickup the success/fail signals automatically.

**Review by ilayn — CHANGES_REQUESTED** [review]
I couldn't find too much time and below are very very quick glance review. I'll come back asap I have some time.

**Comment by ilayn** [comment]
> I was a professional mathematician and software developer for decades, and I understand all the code maintence issues, and have taught code development at the university levelm but I just don't understand GitHub and have NO ONE I can get personal help from. I know no one who uses it as a developer.

Welcome to the club :) Git is ... well let's not waste time on complaining. I know exactly how you feel. It is basically a projection of the mindset of its creator. But on its usage, basically everyone uses it and it is truly the industry standard now. So once you get past of its horrific user interface it is indeed a fantastically powerful tool. The only thing we have to keep in mind is that the user interface is an overloading mess. Same command switch behaves differently if there are other switches change state. Thus requires some frustration and time.

> Here is the result: typical solvers such a lstsq cannot see that they are producing garbage when the system is poorly conditioned, though the matrix is nowhere near being singular.

As you know if the system is square, matlab basically doesn't try to switch to QR solvers it basically gives up with a warning/error in the face of singularity or ill-condition cases. For `lstsq` the story is different as you very well know, this is to do with which LAPACK driver we call and with which parameters. Hence the reason why most behave similarly is because they call the same routine including matlab, scipy, mathematica etc. and none rolls out their own solvers due to industry standards/expectations/conventions.

**Comment by rgommers** [comment]
@ilayn you posted code and a diagram recently for an equivalent function to the Matlab backslash operator. From the description of this PR, `autosol` is most similar to that it seems (right?).

It seems like a similar higher-level diagram for this PR could be very useful.

**Comment by rejones7** [comment]
If I can provide a better explanation please let me know.
As I explained in the (overly long) explanation, this package builds successively on itself:

arlsnn()
     |
arlsgt()
    |
arlseq()
    |
  arls()


________________________________

**Comment by rejones7** [comment]
Yes, I agree that some such merge might be best.
I don't mind a name change either.

But I just really don't understand what you mean by "silent exception" in the case of arls().
Arls() is specifically intended to handle ANY degree of ill-behavior, singularity, etc - on the part of A and b, with absolutely no problem.
I have no idea what I would diagnose to the user.

To borrow a title from the mystery writer, Raymond Chandler, "Trouble is My Business!".
"Give arls() your hardest, your most ill-behaved, your most singular problems! Please! Arls() eats them for breakfast!"

I hope you can take a couple of minutes and look at this example I prepared for you managers of scipy...
http://www.rejones7.net/Autosol/DemoArlsVsLsmr/index.htm

Thanks for your interest and time! I know what it is to be overwhelmed at work! (and home)
Take care.
Ron

________________________________

**Comment by ilayn** [comment]
Ah yes, what I mean by silent exception is that things happen without letting user know about it. Moreover, if an error occurs it basically decides internally with no way of knowing what happened. In other words, 

> it knows what difficulties it is dealing with

but we don't. Hence if the user doesn't want this much smartness it shouldn't try to outsmart the user. But of course there is a balance to strike, not everything can be left for user choices to avoid the function becoming a flight cockpit. So I'm exaggerating a bit. 

And probably still under the influence of your early explanation about "never returning error" behavior.

**Comment by rejones7** [comment]
A day or two ago I uploaded a later version of autosol, and did not document it here. Sorry.

First, there was objection to the name autosol(), so I changed it to arls(),
meaning Automatically Regularized Least Squares.
(Turns out that was a problem too, which I will address a little later.)

Arls.py responds to requests the arls() give the user some feedback about the solution process, as it was "silent".

Each solver now returns:
-- the tradition numerical rank of A ( I checked other current solvers in scipy to be sure I was doing this in the same exact way)
-- the "usable rank" which involves both A and b, and is the crucial artifact needed to do autoregularization
-- the estimate of the right-hand-side RMS error, sigma.
-- the resulting Tikhonov regularization parameter, lambda

(The in-code comments have been updated.)

And, The test_arls() code has been updated to use one function per test, as best as I understand how to do it at this point

By the way, I am not seeking to REPLACE any current scipy solvers. 
Arls contains new techniques and should only be an additional feature, I think.

Rondall Jones

**Comment by rejones7** [comment]
When I switched my package's name from autoreg to arls, I forgot to check it on Google.
So it was pointed out to me that the name "arls" already has a use as a least-squares solver.

Possible names that do NOT have significant collisions on Google include arles, aregls, auregls,
or names like lsmr, including lsar and lsareg. (But "lsa" itself has lots of matches on Google.)

Of all these I think I prefer "arles", as Automatically Regularized LEast Squares. It has no matches except
the town in France where Van Gogh lived.

If there is no particular objection to "arles" I will switch to that.

**Comment by rlucas7** [comment]
I like the arles name and I’m not aware of any least squares based algorithm that conflicts. 

-Lucas

**Comment by WarrenWeckesser** [comment]
I can bikeshed with the best of them...

There are several object names in SciPy that include some sort of abbrevation for "least squares" in the name:

```
interpolate.LSQBivariateSpline
interpolate.LSQSphereBivariateSpline
interpolate.LSQUnivariateSpline
interpolate.make_lsq_spline

linalg.lstsq

optimize.least_squares
optimize.leastsq
optimize.lsq_linear
optimize.nnls

signal.firls

sparse.linalg.lsmr
sparse.linalg.lsqr
```
So we already have `least_squares`, `leastsq`, `lstsq`, `lsq` and `ls`.  It would be nice to pick one those rather than introduce another acronym.  (Personally, I would never guess that `les` has anything to do with least squares.)

**Comment by rejones7** [comment]
Wow, these comments are from 2 years ago!

I finally just uploaded the final product to the contributed library as arls.
it installs with "pip install arls".
It has been downloaded about once a day (not counting mirrors) since I contributed it in 2021.
There have been no requests for changes, bugs, or enhancements.

If you are interested in upgrading it to scipy I would be interested.

Ron Jones
________________________________

**Comment by lucascolley** [comment]
> I finally just uploaded the final product to the contributed library as arls.
it installs with "pip install arls".

Is the source on GitHub somewhere?

**Comment by rejones7** [comment]
I haven’t touched done any python for a couple of years. Please explain more about this to me.
Ron
Sent from my iPhone

On Jan 16, 2024, at 7:21 AM, Lucas Colley ***@***.***> wrote:

﻿

I finally just uploaded the final product to the contributed library as arls.
it installs with "pip install arls".

Is the source on GitHub somewhere?

**Comment by rejones7** [comment]
I finally realized what you were asking.
Yes the source is on Github. But if you install it you get the source, so I don’t understand what you need.
Sent from my iPhone

On Jan 16, 2024, at 8:50 AM, rondall jones ***@***.***> wrote:

﻿ I haven’t touched done any python for a couple of years. Please explain more about this to me.
Ron
Sent from my iPhone

On Jan 16, 2024, at 7:21 AM, Lucas Colley ***@***.***> wrote:

﻿

I finally just uploaded the final product to the contributed library as arls.
it installs with "pip install arls".

Is the source on GitHub somewhere?

**Comment by rejones7** [comment]
My GitHub account is just "rejones7".
I have forgotten how to use github so I hope that helps.

But actually I would not recommend you look at that version.
After years of revamping this method I was able to develop a much much shorter version which is theoretically and practically better, but it is only on MATLAB right now. Would you be interested in converting it?
I will attach the code.
It uses a lot of vectorized coding, which may be a barrier.
There was a write up about this new method here:
https://blogs.mathworks.com/cleve/2023/06/16/arls-automatically-regularized-least-squares/
[http://blogs.mathworks.com/cleve/files/arls_0.bmp]<https://blogs.mathworks.com/cleve/2023/06/16/arls-automatically-regularized-least-squares/>
ARLS, Automatically Regularized Least Squares<https://blogs.mathworks.com/cleve/2023/06/16/arls-automatically-regularized-least-squares/>
(I have a guest blogger today. Ron Jones worked with me in 1985 for his Ph. D. from the University of New Mexico. He retired recently after nearly 40 years at Sandia National Labs in Albuquerque and now has a chance to return to the problem he studied in his thesis. -- CBM)by Rondall Jones, ***@***.*** interest is
blogs.mathworks.com

________________________________

**Comment by lucascolley** [comment]
Thanks for the explanation. Usually we can't take from MATLAB code due to incompatible licenses, but since your source code has its own 3-clause BSD license, I think it would be fine to include in SciPy (I may be wrong).

I'm not personally interested in converting it, but it sounds like that would be the best thing to do if someone wants to add these features.

In light of what I've said, I think it makes sense to close this PR (which is now outdated) and open an issue, so that anyone who would like to work on this can find your MATLAB code. If that sounds wrong to you, let me know!

---

As an aside, I looked on your `rejones7` GitHub account, but it doesn't have the source for `arls`. The only public repositories are a fork of NumPy and a fork of SciPy.

**Pull request closed without merging**